proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,429 @@
+  statements remain at the top of the file.
+
+* Other than corner cases around ambiguity, this proposal is backward compatible; it is not "fork-like".
+
+* Proposal `#160`_ allows users to suppress field selectors, thus ameliorating a small part
+  of what has motivated this proposal.
+
+Costs and Drawbacks
+-------------------
+
+* This is a significant new bit of implementation and specification, and it should require the
+  requisite level of support from the community to be accepted.
+
+Alternatives
+------------
+
+There are many alternatives. The list below is shamelessly cribbed from `#205`_.
+
+* Proposal `#205`_. That proposal essentially tweaks the ``qualified`` feature to
+  become more flexible and exportable. It has proved hard to digest (from the commentary),
+  though, and solves fewer problems than this proposal. On the other hand, it is likely
+  easier to implement.
+
+  This proposal is inspired by some of the topics that came up in the conversation
+  for `#205`_, and I'm grateful for @deepfire's efforts on that proposal.
+
+* 2005 Coutts, ``as`` in export lists: `<https://mail.haskell.org/pipermail/libraries/2005-March/003390.html>`_ . Salient points:
+  letting modules export other modules' contents qualified with the module name`
+  
+* 2006 Wallace, explicit namespaces for module names: `<https://ghc.haskell.org/trac/ghc/wiki/Commentary/Packages/PackageNamespacesProposal>`_ . Salient points:
+  The declaration import namespace brings into availability the subset of the hierarchy of module names rooted in the package "foo-1.3", at the position ``Data.Foo``
+  
+* 2013 de Castro Lopo, qualified exports: `<https://wiki.haskell.org/GHC/QualifiedModuleExport>`_
+  ``qualified module T`` in export list and is essentially a subset of this proposal.
+
+Future Work
+-----------
+
+I see a few future directions along these lines, but I leave it to others to flesh these out.
+
+1. We can imagine *parameterized local modules*, where all the functions defined therein share
+   a sequence of parameters. This would resurrect the ideas behind `#40`_. This would bring
+   us close to ML-style functors.
+
+2. Haskell currently requires three distinct concepts to coincide: *compilation units* are the
+   chunks that go through the compiler all at once, *source files* are distinct files on disk,
+   and *modules* are groups of related definitions and can define an abstraction barrier.
+
+   This proposal allows modules to become smaller than these other two. By writing a module to
+   collect others, modules can also be larger than the other two (as is true today).
+
+   However, it would be nice to separate the treatment of compilation units and source files,
+   as well. This would allow, for example, the inliner and specializer to make decisions with
+   respect to more definitions (if the compilation unit is larger than the source file).
```

Comment:
It would also allow a much simpler implementation of mutually recursive modules: the modules in a SCC form a compilation unit and are processed together, using the usual fixpoint computation.